### PR TITLE
MAINT: cnofs ivm cleaning routine

### PR DESCRIPTION
**Pull request by jklenzing**
Addresses #38.  Proposed updates to cleaning routine for cnofs_ivm.  Tagging @aburrell

### CODE REVIEW

**Comment by aburrell** [comment]
Still not working right, but now in a different way!  The velocity cleaning at the 'clean' level is too permissive.  Using the same time period as #38.

pysatNASA_cnofs_clean_bad

File: pysatNASA/instruments/cnofs_ivm.py [hunk 1]
```diff
@@ -130,12 +130,8 @@ def clean(self):
...
         max_dm_flag = 6
 
-    # First pass, keep good RPA fits
-    idx, = np.where(self.data.RPAflag <= max_rpa_flag)
-    self.data = self[idx, :]
-
-    # Second pass, find bad drifts, replace with NaNs
-    idx = (self.data.driftMeterflag > max_dm_flag)
+    idx = ((self.data.driftMeterflag > max_dm_flag)
+           & (self.data.RPAflag > max_rpa_flag))
```
Comment (jklenzing) [inline pysatNASA/instruments/cnofs_ivm.py:134]:
```suggestion
    idx = ((self.data.driftMeterflag > max_dm_flag)
           | (self.data.RPAflag > max_rpa_flag))
```
I think I found it @aburrell.  We were excluding points where both data was bad, not either data set.  This explains the "too permissive" effect.  Does this work in your analysis?

**Comment by jklenzing** [comment]
Fixed the bugs in this branch.  The plot above won't show what is working in the clean, as the flags are unchanged.  Rather than dropping all data if any set has a problem, the appropriate values are replaced with NaNs.

File: pysatNASA/instruments/cnofs_ivm.py (hunk 1 above)
Comment (jklenzing) [reply pysatNASA/instruments/cnofs_ivm.py]: replaced by later updates

**Comment by aburrell** [comment]
Still too strict for density, I think:

ionDensity_1112021_clean_levels
ionTemperature_1112021_clean_levels
ionVelparallel_1112021_clean_levels

**Comment by aburrell** [comment]
Additionally, it looks like there are 2 ion density variables?  (ionDensity, Ni).

**Comment by jklenzing** [comment]
> Additionally, it looks like there are 2 ion density variables? (ionDensity, Ni).

From the documentation, it looks like `ionDensity` is from the RPA fit,`Ni` from the zero-volt current.  I'm not sure if we should filter out both of these are not.

**Comment by aburrell** [comment]
Since it's a clean routine for the data set, it seems like we should be filtering all of the data appropriately.

**Comment by jklenzing** [comment]
Found it.  RPA flag values of 4 should pass through always for density.  Since the raw files remove values higher than this anyway, I just pulled out this whole if statement.

**Comment by aburrell** [comment]
@jklenzing let me know when you've made all the changes we discussed at the meeting.

**Comment by jklenzing** [comment]
I've pushed the latest changes, but have not had a chance to make plots

**Comment by aburrell** [comment]
Poke @jklenzing 🐛 🐞

**Comment by jklenzing** [comment]
cindi

**Comment by jklenzing** [comment]
Since we're not dropping flags anymore, can't break these out as in the figures above.

**Comment by aburrell** [comment]
Ok, I noticed some things we had talked about the last time we looked at this together that weren't changed, so I just went in and changed the code myself.  This now works as expected and raises less warnings.

ionTemp_test

ionParallelVel_test

**Comment by aburrell** [comment]
Requested review from @JonathonMSmith since I changed enough that I shouldn't be allowed to approve.

**Comment by JonathonMSmith** [comment]
I think this looks like it's working as expected. Looking at 'clean' data here.

flag_1
Flag2


This is definitely beyond the scope of this pull and maybe a bad idea altogether. As I'm looking at the cleaning routine I keep wondering, would it be possible to break out the masking routines as functions that return the indices for debugging purposes? I wonder if it might help to understand which nans come from which mask.

**Review by JonathonMSmith — APPROVED** [review]
Really appreciate the improved comments for each of the cleaning processes

**Comment by aburrell** [comment]
> This is definitely beyond the scope of this pull and maybe a bad idea altogether. As I'm looking at the cleaning routine I keep wondering, would it be possible to break out the masking routines as functions that return the indices for debugging purposes? I wonder if it might help to understand which nans come from which mask

I think this a good idea that could help new users understand what is going on and why.  The cleaning routine could be broken up into different `instruments.methods.cnofs` routines such as `clean_velocity`, `clean_dens`, `clean_temp`, and `clean_time`.

**Pull request merged**